sql-plugin/src/main/scala/com/nvidia/spark/rapids/conditionalExpressions.scala

```diff
@@ -129,13 +159,39 @@ case class GpuCaseWhen(
     }
   }
 
-  override def columnarEval(batch: ColumnarBatch): Any = {
-    // `elseRet` will be closed in `computeIfElse`.
-    val elseRet = elseValue
-      .map(_.columnarEval(batch))
-      .getOrElse(GpuScalar(null, branches.last._2.dataType))
-    branches.foldRight[Any](elseRet) { case ((predicateExpr, trueExpr), falseRet) =>
-      computeIfElse(batch, predicateExpr, trueExpr, falseRet)
+  @transient
+  private[this] lazy val trueExpressions = branches.map(_._2)
+
+  override def columnarEval(batch:ColumnarBatch): Any = {
+    val size = branches.size
+    val predictions = new Array[GpuColumnVector](size)
+    var isAllPredsFalse = true
+    var i = 0
+
+    withResource(predictions) { preds =>
+      while (i < size) {
+        // If any predication is the first all-true, then evaluate its true expression
+        // and return the result.
+        preds(i) = GpuExpressionsUtils.columnarEvalToColumn(branches(i)._1, batch)
+        val p = preds(i)
```

Comment:
Fine with me. Updated.
I just wanted to put it into the cache as soon as possible.

Reply:
I just wanted to put the result into the cache as soon as possible to avoid leaking, even the assignment `preds(i) = p` is almost impossible to fail. But I am find with the suggestion. Updated.